### PR TITLE
SFB-115: EURO input field

### PR DESCRIPTION
**Pull request by JonasVanHoof**
## ID
 [SFB-115](https://binnenland.atlassian.net/browse/SFB-115)

 ## Description

For Subsidie forms, an amount in euros is often asked.

Today, we use a numeric input field for that, but then there is no formatting, which can lead to inconvenience.

For example, 1000000 → very unclear. You have to count the zeros to know what amount we are talking about

 ## Type of change

 - [ ] Bug fix
 - [x] New feature
 - [ ] Breaking change
 - [ ] Other

 ## How to test
Go to the dummy app and try the new currency input field.

 ## Links to other PR's
- /

## Attachments

V1
image

V2
image

### CODE REVIEW

**Comment by JonasVanHoof** [comment]
The component is create with a static euro sign and the default decimal and thousand separator. Should we add predicates to the input field so we can set those from our ttl? Or is this enough for our usecase?

@aatauil

File: addon/components/rdf-input-fields/currency-input.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  inputId = 'input-' + guidFor(this);
+
+  currencySymbol = '€';
+
+  @action
+  updateValue(e) {
+    if (e && typeof e.preventDefault === 'function') e.preventDefault();
+    this.value = e.target.value.trim();
+
+    super.updateValue(this.value);
```
Comment (Windvis) [inline addon/components/rdf-input-fields/currency-input.js:15]: Question: 1 thing I'm not sure about is the data type. At the moment this will always be a string, but I'm not sure if that's the best type for currencies in RDF, but I'm no expert on that. Ignore this if string is the way to go.

File: addon/components/rdf-input-fields/currency-input.hbs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  >
+    {{@field.label}}
+  </AuLabel>
+{{/unless}}
+
+<AuButtonGroup @inline={{true}}>
+  <AuButton
+    @skin="primary"
+    @disabled={{true}}
+  >{{this.currencySymbol}}</AuButton>
```
Comment (Windvis) [inline addon/components/rdf-input-fields/currency-input.hbs:16]: Creative solution, but this reminds me that we need to add an ["input group" feature](https://github.com/appuniversum/ember-appuniversum/issues/368) to Appuniversum 😄, or some way to display custom content instead of only the icons we support now.

I noticed that the cursor will show a 🚫 icon when hovering the euro sign, because it's a disabled button. I wonder if we should add some style overrides to the addon for this specific case, until Appuniversum supports input groups? (`pointer-events: none` most likely?)

**Review by Windvis — COMMENTED** [review]
Some small remarks / questions but it looks good to me 👍.

**Comment by Windvis** [comment]
> The component is create with a static euro sign and the default decimal and thousand separator. Should we add predicates to the input field so we can set those from our ttl? Or is this enough for our usecase?

I think it's good as-is. I doubt we would need other currencies in the future, but if we do we can add the options then as well.

File: addon/components/rdf-input-fields/currency-input.js (hunk 1 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.js:15]: @aatauil  

Good point. Had some though about it to. But not sure how they want the value in the database. They will always build up the currency symbol with the value I assume? Didn't get an example on how this reflects to the old use in the forms.

File: addon/components/rdf-input-fields/currency-input.hbs (hunk 2 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.hbs]: Unfortunately this didn't work with the pointer events none.. Also tried the `cursor: pointer !important` but it did not work..

As t is a new component i create a class for the currencyInput component. Maybe see if we want this custom css or just the disabled cursor event

File: addon/styles/addon.css [hunk 3]
```diff
@@ -38,3 +38,33 @@
...
+    border-bottom-right-radius: 0;
+    margin-right: -0.5px;
+  }
+
+  input {
+    border-top-left-radius: 0 !important;
+    border-bottom-left-radius: 0 !important;
+    border-left: none !important;
+  }
+}
```
Comment (JonasVanHoof) [inline addon/styles/addon.css:70]: Tried using the app universum variables but did not work for everything

File: addon/components/rdf-input-fields/currency-input.hbs (hunk 2 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.hbs]: image

File: addon/components/rdf-input-fields/currency-input.js (hunk 1 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.js:15]: We could add some more values for it but maybe that's overkill? the data structure would look like this
```ttl
# Currency Input
<http://ember-submission-form-fields/source-node>
  <http://mu.semte.ch/vocabularies/ext/currencyInputValue> "0,0" ;
  <http://schema.org/currency> "€" ;
  <http://schema.org/price> "€0,0" ;
  <http://schema.org/amount> 0.0 .

File: addon/components/rdf-input-fields/currency-input.hbs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    id={{this.inputId}}
+    value={{this.value}}
+    {{on "blur" this.updateValue}}
+    {{au-inputmask
+      options=(hash
+        alias="currency"
+        radixPoint=","
+        groupSeparator="."
+        allowMinus=true
+        inputType="decimal"
```
Comment (Windvis) [inline addon/components/rdf-input-fields/currency-input.hbs:27]: I took a quick look at the docs, and this doesn't seem to be mentioned as an option? https://robinherbots.github.io/Inputmask/#/documentation/numeric#inputtype

Maybe I'm missing something, the docs are a bit vague.

File: addon/components/rdf-input-fields/currency-input.hbs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      )
+    }}
+  />
+  <div class="currencyIsoCode">{{this.currencyIsocode}}</div>
```
Comment (Windvis) [inline addon/components/rdf-input-fields/currency-input.hbs:31]: Question: Why was this changed? Is this the better way to display the euro currency or something?

File: addon/styles/addon.css [hunk 6]
```diff
@@ -38,3 +38,32 @@
 .concept-scheme-multi-select-checkboxes__content {
   width: 100%;
 }
+
+/* SFB-115 */
```
Comment (Windvis) [inline addon/styles/addon.css:42]: Nitpick, but this Jira id has little value as a comment IMO 😄, it's probably already mentioned in the PR where this commit is linked to.

File: addon/styles/addon.css [hunk 7]
```diff
@@ -38,3 +38,32 @@
...
+.CurrencyInput {
```
Comment (Windvis) [inline addon/styles/addon.css:43]: Nitpick: I would use BEM casing to be consistent to how Appuniversum names things. `.currency-input`.

File: addon/styles/addon.css [hunk 8]
```diff
@@ -38,3 +38,32 @@
...
+  display: flex;
+  flex-direction: row nowrap;
+  gap: 0;
+  align-items: center;
+
+  .currencyIsoCode {
```
Comment (Windvis) [inline addon/styles/addon.css:49]: `.currency-input__iso-code`

File: addon/components/rdf-input-fields/currency-input.js (hunk 1 above)
Comment (Windvis) [reply addon/components/rdf-input-fields/currency-input.js:15]: Sorry, the conclusion is a bit vague for me. String is the way to go? The chat seemed to imply that numbers would be preferred for sorting / filtering reasons?

File: addon/components/rdf-input-fields/currency-input.hbs (hunk 2 above)
Comment (Windvis) [reply addon/components/rdf-input-fields/currency-input.hbs]: > Unfortunately this didn't work with the pointer events none.. Also tried the cursor: pointer !important but it did not work..

Strange, I tried that in my dev tools 😄 maybe it works in Firefox. Anyway, custom styles is nice as well 👍.

File: addon/components/rdf-input-fields/currency-input.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+export default class RdfInputFieldsCurrencyInputComponent extends SimpleInputFieldComponent {
+  inputId = 'input-' + guidFor(this);
+
+  currencyIsocode = 'EUR';
+
+  @action
+  updateValue(e) {
+    if (e && typeof e.preventDefault === 'function') e.preventDefault();
+    this.value = e.target.value.trim();
```
Comment (Windvis) [inline addon/components/rdf-input-fields/currency-input.js:13]: [inputmask also adds a `inputmask` property to the target which has an ](https://robinherbots.github.io/Inputmask/#/documentation#unmaskedvalue)`unmaskedvalue()` method. I'm not sure if we want the mask to be stored in the DB?

File: addon/components/rdf-input-fields/currency-input.hbs (hunk 4 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.hbs:27]: I was very confused yesterday with the `inputmode` or `inputType` in the examples but now it's clear. You are right. I removed the `inputType` as the default is already `text` and `inputmode` is `decimal` as default so that can also stay.

File: addon/styles/addon.css (hunk 8 above)
Comment (JonasVanHoof) [reply addon/styles/addon.css]: Updated

File: addon/styles/addon.css (hunk 7 above)
Comment (JonasVanHoof) [reply addon/styles/addon.css]: Removed comment + use of BEM. Only not sure if I can leave the `input` inside the currency-input class. If we follow the BEM I think we should add a specific class or id to it

File: addon/components/rdf-input-fields/currency-input.hbs (hunk 5 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.hbs]: The currency symbols could be non-alphanumeric characters. I think we will not be needing other currencies but it keeps us away from any issues with the symbols. The iso code is a standard and can be used for every currency. It has a 3 letter format and is more readable than `€`

`€` or `EUR`
`$` or `USD`

I think if you read this you are also sold on using this :)
`ISO 4217 is a standard published by the International Organization for Standardization that defines alpha codes and numeric codes for the representation of currencies and provides information about the relationships between individual currencies and their minor units.`

File: addon/components/rdf-input-fields/currency-input.js (hunk 9 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.js:13]: I think you should store the mask in the database.. We will get a data model like this... and the consumer of the data can do everything with this (calculations, viewing, sorting,..). But the thing here is as we can only define ONE sh:path on a field we can only do the string input value with that. We can of course just add the other predicates to the same subject in the ember component itself. Do we want this, is this allowed in ESFF? 
```ttl
# Currency Input
<http://ember-submission-form-fields/source-node>
  <http://mu.semte.ch/vocabularies/ext/currencyInputValue> "0,0" ;
  <http://schema.org/currency> "EUR" ;
  <http://schema.org/price> "0,0 EUR" ;
  <http://schema.org/amount> 0.0 .
```

=> the reason why we did not save the number already is that we have to convert the incoming and outgoing value each time. as the input-mask cannot handle a number `123456.00` to be displayed as `123456,00` as we have set the radix point to a comma and not the dot. It will be shown as `12345600,00`. Another use case that the currency,price and amount predicates will be handy for.

File: addon/components/rdf-input-fields/currency-input.js (hunk 1 above)
Comment (aatauil) [reply addon/components/rdf-input-fields/currency-input.js:15]: @JonasVanHoof Conclusion here is to just store the value as a decimal in the database ? from what I understand in rocketchat

File: addon/components/rdf-input-fields/currency-input.js (hunk 9 above)
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.js:13]: Will just save it as a decimal and convert the value each time

File: tests/dummy/public/test-forms/basic-fields/data.ttl [hunk 10]
```diff
@@ -19,6 +22,13 @@
 # Input
 <http://ember-submission-form-fields/source-node> <http://mu.semte.ch/vocabularies/ext/inputValue> "Simple value".
 
+# Currency Input
+esff:source-node ext:currencyInputValue eurio:currencyInput.
+
+eurio:currencyInput a eurio:MonetaryAmount ;
+    eurio:value "10000.00"^^xsd:decimal ;
+    eurio:currency "EUR" .
+
```
Comment (JonasVanHoof) [inline tests/dummy/public/test-forms/basic-fields/data.ttl:31]: The table in the dummy app that shows the triples:

image
image

Comment (JonasVanHoof) [reply tests/dummy/public/test-forms/basic-fields/data.ttl:31]: The value is typed as an integer as the fractions are 00

File: addon/components/rdf-input-fields/currency-input.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  @action
+  updateValue(e) {
+    if (e && typeof e.preventDefault === 'function') e.preventDefault();
+
+    if (e.target && e.target.value) {
+      const amount = stringAmountToNumber(e.target.value, {
+        decimalSeparator: this.decimalSeparator,
+        thousandSeparator: this.thousandSeparator,
+      });
+      this.value = this.numberStringToFormattedString(amount);
```
[the remark this answers is not in the document]
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.js:22]: I cannot get it to work with your setup. Did you change the decimals in the input field or also the fractions? I noticed that it does work if you only change the decimals and not the fractions. 

When reading the docs it also seems to be not possible to have an `inputType="number"` with a `radixPoint = ","`
image

[the remark this answers is not in the document]
Comment (Windvis) [reply addon/components/rdf-input-fields/currency-input.js:22]: I only changed the inputType value, and made sure that `this.value` was set to the number I got from `e.target.inputmask.unmaskedvalue()`. I'll create a PR with the changes so we're looking at the same code.

[the remark this answers is not in the document]
Comment (Windvis) [reply addon/components/rdf-input-fields/currency-input.js:22]: This is the code that _seems_ to be working on first glance: https://github.com/lblod/ember-submission-form-fields/pull/176

[the remark this answers is not in the document]
Comment (Windvis) [reply addon/components/rdf-input-fields/currency-input.js:22]: > When reading the docs it also seems to be not possible to have an inputType="number" with a radixPoint = ","

I _think_ this is talking about the value that you `.set` on the form field, so not the values that users enter by typing. That's how I understood it at least. So if you set `inputType` to number, you need to provide a number instead of the string when using `.set` (which Ember does when you pass it as the `value={{this.value}}` attribute.

[the remark this answers is not in the document]
Comment (Windvis) [reply addon/components/rdf-input-fields/currency-input.js:22]: Strange, it seems we _are_ still passing a string because of the `formattedValue` getter, but it still worked on my machine 😄. Changing it to `this.value` directly also seems to work. 🤔 

Let me try in Chrome.

[the remark this answers is not in the document]
Comment (JonasVanHoof) [reply addon/components/rdf-input-fields/currency-input.js:22]: > Strange, it seems we _are_ still passing a string because of the `formattedValue` getter, but it still worked on my machine 😄. Changing it to `this.value` directly also seems to work. 🤔
> 
> Let me try in Chrome.
> 
> Edit, also works in Chrome.

Did you just change the input value to this.value instead of the getter get formattedValue?

**Review by Windvis — APPROVED** [review]
